core/drivers/crypto/versal/rsa.c

```diff
@@ -0,0 +1,364 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2022.
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <drvcrypt.h>
+#include <drvcrypt_acipher.h>
+#include <crypto/crypto_impl.h>
+#include <initcall.h>
+#include <ipi.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+#include <tee/cache.h>
+
+#define RSA_MAX_PRIV_EXP_LEN	512
+#define RSA_MAX_PUB_EXP_LEN	4
+#define RSA_MAX_MOD_LEN		512
+
+static void crypto_bignum_bn2bin_pad(size_t size,
+				     struct bignum *from, uint8_t *to)
+{
+	size_t len = crypto_bignum_num_bytes(from);
+
+	crypto_bignum_bn2bin(from, to + size - len);
+}
+
+static TEE_Result do_encrypt(struct drvcrypt_rsa_ed *rsa_data)
+{
+	struct rsa_public_key *p = rsa_data->key.key;
+	struct versal_rsa_input_param *cmd = NULL;
+	struct versal_mbox_mem cmd_buf = { };
+	struct versal_mbox_mem cipher = { };
+	struct versal_mbox_mem key = { };
+	struct versal_mbox_mem msg = { };
+	struct versal_cmd_args arg = { };
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t err = 0;
+
+	switch (rsa_data->rsa_id) {
+	case DRVCRYPT_RSA_PKCS_V1_5:
+		if (rsa_data->key.n_size != 128)
+			break;
+
+		return sw_crypto_acipher_rsaes_encrypt(rsa_data->algo,
+						rsa_data->key.key,
+						rsa_data->label.data,
+						rsa_data->label.length,
+						rsa_data->message.data,
+						rsa_data->message.length,
+						rsa_data->cipher.data,
+						&rsa_data->cipher.length);
+		break;
+	case DRVCRYPT_RSA_OAEP:
+		return sw_crypto_acipher_rsaes_encrypt(rsa_data->algo,
+						rsa_data->key.key,
+						rsa_data->label.data,
+						rsa_data->label.length,
+						rsa_data->message.data,
+						rsa_data->message.length,
+						rsa_data->cipher.data,
+						&rsa_data->cipher.length);
+		break;
+	case DRVCRYPT_RSA_NOPAD:
+		return sw_crypto_acipher_rsanopad_encrypt(rsa_data->key.key,
+						rsa_data->message.data,
+						rsa_data->message.length,
+						rsa_data->cipher.data,
+						&rsa_data->cipher.length);
+		break;
+	case DRVCRYPT_RSASSA_PKCS_V1_5:
+		if (rsa_data->hash_algo == TEE_ALG_SHA1 ||
+		    rsa_data->key.n_size == 128)
+			panic();
```

Comment:
`panic()` should be reserved for cases when it is impossible to continue safely.

Reply:
right. if this code was executed it means that someone tampered or glitched the execution....it should be impossible to reach it otherwise.

Reply:
"impossible" conditions should normally be checked with `assert()`.

Reply:
you are right, will update